mitiq/pec/channels.py

```diff
@@ -0,0 +1,219 @@
+    H,
+    CNOT,
+    LineQubit,
+    DensityMatrixSimulator,
+    DepolarizingChannel,
+    AmplitudeDampingChannel,
+    channel,
+)
+
+
+def _max_ent_state_circuit(num_qubits: int) -> Circuit:
+    r"""Generates a circuit which prepares the maximally entangled state
+    |\omega\rangle = U |0\rangle  = \sum_i |i\rangle \otimes |i\rangle .
+
+    Args:
+        num_qubits: The number of qubits on which the circuit is applied.
+            It must be an even number because of the structure of a
+            maximally entangled state.
+
+    Returns:
+        The circuits which prepares the state |\omega\rangle.
+
+    Raises:
+        Value error: if num_qubits is not an even positive integer.
+    """
+
+    if not isinstance(num_qubits, int) or num_qubits % 2 or num_qubits == 0:
+        raise ValueError(
+            "The argument 'num_qubits' must be an even and positive integer."
+        )
+
+    alice_reg = LineQubit.range(num_qubits // 2)
+    bob_reg = LineQubit.range(num_qubits // 2, num_qubits)
+
+    return Circuit(
+        # Prepare alice_register in a uniform superposition
+        H.on_each(*alice_reg),
+        # Correlate alice_register with bob_register
+        [CNOT.on(alice_reg[i], bob_reg[i]) for i in range(num_qubits // 2)],
+    )
+
+
+def _circuit_to_choi(circuit: Circuit) -> np.ndarray:
+    """Returns the density matrix of the Choi state associated to the
+    input circuit.
+
+    The density matrix completely characterizes the quantum channel induced by
+    the input circuit (including the effect of noise if present).
+
+    Args:
+        circuit: The input circuit.
+    Returns:
+        The density matrix of the Choi state associated to the input circuit.
+    """
+    simulator = DensityMatrixSimulator()
+    num_qubits = len(circuit.all_qubits())
+    # Copy and remove all operations
+    full_circ = deepcopy(circuit)[0:0]
+    full_circ += _max_ent_state_circuit(2 * num_qubits)
+    full_circ += circuit
+    return simulator.simulate(full_circ).final_density_matrix  # type: ignore
+
+
+def _operation_to_choi(operation_tree: OP_TREE) -> np.ndarray:
+    """Returns the density matrix of the Choi state associated to the
+    input operation tree (e.g. a single operation or a sequence of operations).
+
+    The density matrix completely characterizes the quantum channel induced by
+    the input operation tree (including the effect of noise if present).
+
+    Args:
+        operation_tree: Nested list of operations.
+    Returns:
+        The density matrix of the Choi state associated to the input circuit.
+    """
+    circuit = Circuit(operation_tree)
+    return _circuit_to_choi(circuit)
+
+
+def _tensor_product_kraus(
+    local_kraus: List[np.ndarray], num_channels: int
+) -> List[np.ndarray]:
+    """Given the a local channel (defined by its kraus operators),
+    this functions returns the Kraus operators corresponding to tensor
+    product of "num_channels" copies of the input channel.
+    """
```

Comment:
I think the docstring / arguments could be clearer here. Maybe `nqubits` or `ncopies` instead of `num_channels`? Also, there's no requirement that `local_kraus` has to be local really, and local isn't defined. 

I guess what you're doing is:

Given Kraus operators K_1, ..., K_m, returns {K_{i_1} \otimes \cdots \otimes K_{i_m}} for all m^`ncopies` combinations (i_1, ..., i_m) of indices (1, ..., m).

Reply:
> Given Kraus operators K_1, ..., K_m, returns {K_{i_1} \otimes \cdots \otimes K_{i_m}} for all m^`ncopies` combinations (i_1, ..., i_m) of indices (1, ..., m).
Yes, this is correct. The resulting kraus operators correspond to the tensor product of `ncopies` of the input channel.

Anyway, I removed this function since (now that we have `tensor_product()`), it can be directly written with a single line:
`[tensor_product(*tup) for tup in product(local_kraus, repeat=num_qubits)]`